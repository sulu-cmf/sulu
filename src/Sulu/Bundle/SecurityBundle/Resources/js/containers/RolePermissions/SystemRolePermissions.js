// @flow
import React from 'react';
import {action, computed, observable} from 'mobx';
import {observer} from 'mobx-react';
import {Heading, Matrix, Toggler} from 'sulu-admin-bundle/components';
import {translate} from 'sulu-admin-bundle/utils';
import securityContextStore from '../../stores/securityContextStore';
import {getActionIcon} from '../../utils/Permission';
import type {Role} from '../../types';
import systemRolePermissionsStyles from './systemRolePermissions.scss';
import type {RolePermissions} from './types';

type Props = {|
    actions: Array<string>,
    disabled: boolean,
    onChange: (value: RolePermissions, system: string) => void,
    resourceKey: string,
    roles: Array<Role>,
    system: string,
    values: RolePermissions,
|};

@observer
class SystemRolePermissions extends React.Component<Props> {
    @observable active: boolean = false;

    @action componentDidMount() {
        this.active = this.hasValues;
    }

    handleChange = (values: RolePermissions) => {
        const {onChange, system} = this.props;
        onChange(values, system);
    };

    @action handleActiveChange = (active: boolean) => {
        this.active = active;

        if (!this.active) {
            const {onChange, system} = this.props;
            onChange({}, system);
        }
    };

    @computed get defaultValue() {
        const {resourceKey, roles} = this.props;

        if (!roles) {
            return {};
        }

        const securityContext = securityContextStore.getSecurityContextByResourceKey(resourceKey);

        return roles.reduce((value, role) => {
            const rolePermission = role.permissions.find((permission) => permission.context === securityContext);
            value[role.id] = securityContextStore.getAvailableActions(resourceKey, role.system)
                .reduce((actionValue, action) => {
                    actionValue[action] = rolePermission ? rolePermission.permissions[action] : false;

                    return actionValue;
                }, {});

            return value;
        }, {});
    }

    @computed get hasValues() {
        const {values} = this.props;

        return Object.keys(values).length > 0;
    }

    render() {
        const {actions, disabled, roles, system, values} = this.props;

        return (
            <div className={systemRolePermissionsStyles.systemRolePermissions}>
                <Heading label={translate('sulu_security.system_permission_heading', {system})}>
                    <Toggler checked={this.active} onChange={this.handleActiveChange} />
                </Heading>
                {this.active &&
                    <Matrix
                        className={systemRolePermissionsStyles.matrix}
                        disabled={disabled}
                        onChange={this.handleChange}
                        values={this.hasValues ? values : this.defaultValue}
                    >
                        {roles.map((role) => (
                            <Matrix.Row key={role.id} name={role.id.toString()} title={role.name}>
                                {actions.map((action) => (
                                    <Matrix.Item icon={getActionIcon(action)} key={action} name={action} />
                                ))}
                            </Matrix.Row>
                        ))}
                    </Matrix>
                }
            </div>
        );
    }
}

export default SystemRolePermissions;
