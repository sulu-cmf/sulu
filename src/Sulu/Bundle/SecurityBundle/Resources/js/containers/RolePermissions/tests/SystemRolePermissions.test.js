// @flow
import React from 'react';
import {shallow, render} from 'enzyme';
import SystemRolePermissions from '../SystemRolePermissions';

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test('Render permissions for a single system', () => {
    const roles = [
        {id: 2, identifier: '', name: 'User', permissions: [], system: 'Sulu'},
        {id: 3, identifier: '', name: 'Contact Manager', permissions: [], system: 'Sulu'},
    ];

    expect(render(
        <SystemRolePermissions
            actions={['view', 'add', 'edit']}
            disabled={false}
            onChange={jest.fn()}
            roles={roles}
            system="Sulu"
            values={{'2': {view: true, add: false, edit: true}, '3': {view: false, add: true, edit: false}}}
        />
    )).toMatchSnapshot();
});

test('Render permissions for a single system in disabled state', () => {
    const systemRolePermissions = shallow(
        <SystemRolePermissions
            actions={[]}
            disabled={true}
            onChange={jest.fn()}
            roles={[]}
            system="Sulu"
            values={{}}
        />
    );

    expect(systemRolePermissions.find('Matrix').prop('disabled')).toEqual(true);
});

test('Call onChange callback when matrix changes', () => {
    const changeSpy = jest.fn();

    const systemRolePermissions = shallow(
        <SystemRolePermissions
            actions={['view']}
            disabled={false}
            onChange={changeSpy}
            roles={[]}
            system="Sulu"
            values={{}}
        />
    );

    const newValue = {'1': {view: true}};
    systemRolePermissions.find('Matrix').simulate('change', newValue);

    expect(changeSpy).toBeCalledWith(newValue);
});
