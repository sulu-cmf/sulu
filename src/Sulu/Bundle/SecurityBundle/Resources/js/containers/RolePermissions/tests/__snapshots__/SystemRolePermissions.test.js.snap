// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render permissions for a single system 1`] = `
<div
  class="systemRolePermissions"
>
  <div
    class="line"
  >
    <div
      class="label"
    >
      sulu_security.system_permission_heading
    </div>
    <label
      class="label"
      tabindex="-1"
    >
      <span
        class="switch toggler"
      >
        <input
          type="checkbox"
          value=""
        />
        <span />
      </span>
    </label>
  </div>
  <table
    class="matrix matrix"
  >
    <tbody>
      <tr
        class="row"
      >
        <td
          class="name"
        >
          User
        </td>
        <td
          class="items"
        >
          <button
            class="item selected"
            title="View"
            type="button"
          >
            <span
              aria-label="su-eye"
              class="su-eye"
            />
          </button>
          <button
            class="item"
            title="Add"
            type="button"
          >
            <span
              aria-label="su-plus-circle"
              class="su-plus-circle"
            />
          </button>
          <button
            class="item selected"
            title="Edit"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="rowButton"
            type="button"
          >
            sulu_admin.deactivate_all
          </button>
        </td>
      </tr>
      <tr
        class="row"
      >
        <td
          class="name"
        >
          Contact Manager
        </td>
        <td
          class="items"
        >
          <button
            class="item"
            title="View"
            type="button"
          >
            <span
              aria-label="su-eye"
              class="su-eye"
            />
          </button>
          <button
            class="item selected"
            title="Add"
            type="button"
          >
            <span
              aria-label="su-plus-circle"
              class="su-plus-circle"
            />
          </button>
          <button
            class="item"
            title="Edit"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="rowButton"
            type="button"
          >
            sulu_admin.deactivate_all
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
